app/src/main/java/org/kiwix/kiwixmobile/local_file_transfer/WifiDirectManager.kt

```diff
@@ -0,0 +1,424 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package org.kiwix.kiwixmobile.local_file_transfer
+
+import android.app.Activity
+import android.content.BroadcastReceiver
+import android.content.Context
+import android.content.IntentFilter
+import android.net.Uri
+import android.net.wifi.WpsInfo
+import android.net.wifi.p2p.WifiP2pConfig
+import android.net.wifi.p2p.WifiP2pDevice
+import android.net.wifi.p2p.WifiP2pDeviceList
+import android.net.wifi.p2p.WifiP2pInfo
+import android.net.wifi.p2p.WifiP2pManager
+import android.net.wifi.p2p.WifiP2pManager.ActionListener
+import android.net.wifi.p2p.WifiP2pManager.Channel
+import android.net.wifi.p2p.WifiP2pManager.ChannelListener
+import android.net.wifi.p2p.WifiP2pManager.ConnectionInfoListener
+import android.net.wifi.p2p.WifiP2pManager.PeerListListener
+import android.os.AsyncTask
+import android.os.Build.VERSION
+import android.os.Build.VERSION_CODES
+import android.os.Looper
+import android.util.Log
+import android.widget.Toast
+import org.kiwix.kiwixmobile.core.BuildConfig
+import org.kiwix.kiwixmobile.core.R.string
+import org.kiwix.kiwixmobile.core.utils.AlertDialogShower
+import org.kiwix.kiwixmobile.core.utils.KiwixDialog.FileTransferConfirmation
+import org.kiwix.kiwixmobile.core.utils.SharedPreferenceUtil
+import org.kiwix.kiwixmobile.local_file_transfer.FileItem.FileStatus
+import org.kiwix.kiwixmobile.local_file_transfer.KiwixWifiP2pBroadcastReceiver.P2pEventListener
+import java.io.IOException
+import java.io.InputStream
+import java.io.OutputStream
+import java.net.InetAddress
+import java.util.ArrayList
+import javax.inject.Inject
+
+/**
+ * Manager for the Wifi-P2p API, used in the local file transfer module
+ */
+class WifiDirectManager @Inject constructor(
+  private val activity: Activity,
+  sharedPreferenceUtil: SharedPreferenceUtil,
+  alertDialogShower: AlertDialogShower
+) : ChannelListener, PeerListListener, ConnectionInfoListener, P2pEventListener {
+  private val callbacks: Callbacks
+  private val sharedPreferenceUtil: SharedPreferenceUtil
+  private val alertDialogShower: AlertDialogShower
+  /* Helper methods */
+  /* Variables related to the WiFi P2P API */
+  var isWifiP2pEnabled = false // Whether WiFi has been enabled or not
+    private set
+  private var shouldRetry =
+    true // Whether channel has retried connecting previously
+  private var manager // Overall manager of Wifi p2p connections for the module
```

Comment:
I am not sure if it is valid for any of these objects but it would be good if any of them could be `lateinit` so that we don't have to declare them nullable